File: trie/prefix-trie.cpp
#include <iostream>
#include <unordered_map>
using namespace std;

class Node {
  public:
    char data;
    unordered_map<char, Node*> map;
    bool terminal;
    Node(char _data){
      data = _data;
      terminal = false;
    }
};

class Trie {
public:
    Node* root;
    Trie() {
      root = new Node('\0');
    }
    
    void insert(string word) {
      Node *cursor = root;
      for(char w : word){
        if(cursor->map.count(w) == 0){
          cursor->map[w] = new Node(w);
        }
        cursor = cursor->map[w];
      }
      cursor->terminal = true;
    }
    
    bool search(string word) {
        Node *cursor = root;
        for(char w : word){
          if(cursor->map.count(w) == 0){
            return false;
          }
          cursor = cursor->map[w];
        }
        return cursor->terminal;
    }
    
    bool startsWith(string prefix) {
      Node *cursor = root;
      for(char w : prefix){
        if(cursor->map.count(w) == 0){
          return false;
        }
        cursor = cursor->map[w];
      }
      return true;
    }
};

int main(){
  Trie* trie = new Trie();
  trie->insert("hello");
  return 0;
}

File: trie/prefix-trie.js


class Node {
  constructor(val, isTerminal = false){
    this.val = val;
    this.map = {};
    this.isTerminal = isTerminal;
  }
}
class Trie {
  constructor(){
    this.root = new Node('\0')
  }

  insert(word){
    let cursor = this.root;
    for(const char of word){
      if(cursor.map[char] === undefined){
        const newNode = new Node(char);
        cursor.map[char] = newNode;
      } 
      cursor = cursor.map[char];
    }
    cursor.isTerminal = true;
  }

  search(word){
    let cursor = this.root;
    for(const char of word){
      if(cursor.map[char] === undefined) {
        return false;
      }
      cursor = cursor.map[char];
    }
    return cursor.isTerminal;
  }
}

const words = ['apple', 'ape', 'no', 'news', 'not', 'never']
const trie = new Trie();
for(const word of words){
  trie.insert(word);
}
console.log(trie.search('apple'))
